packages/babel-plugin-proposal-object-rest-spread/src/index.js

```diff
@@ -131,6 +131,48 @@ export default declare((api, opts) => {
     });
   }
 
+  function createObjRefExpression(path, objRef) {
+    let impureDefaultValueDeclarator;
+    let objRefExpression;
+    if (path.parentPath?.isAssignmentPattern()) {
+      // { { foo, ...bar } = qux } = quux;
+      const { scope } = path;
+      const parentRight = path.parentPath.get("right");
+      const parentRightId = scope.maybeGenerateMemoised(parentRight.node);
+      if (parentRightId) {
+        impureDefaultValueDeclarator = t.variableDeclarator(
+          parentRightId,
+          t.cloneNode(parentRight.node),
+        );
+      }
+      const objRefName = scope.generateUidBasedOnNode(objRef, "_ref");
+      scope.push({ id: t.identifier(objRefName) });
+
+      objRefExpression = t.sequenceExpression([
+        t.assignmentExpression(
+          "=",
+          t.identifier(objRefName),
+          t.cloneNode(objRef),
+        ),
+        t.conditionalExpression(
+          t.binaryExpression(
+            "!==",
+            t.identifier(objRefName),
+            scope.buildUndefinedNode(),
+          ),
+          t.identifier(objRefName),
+          parentRightId ?? t.cloneNode(parentRight.node),
```

Comment:
It `parentRightId` is null, won't this generate an invalid AST? AFAIK the third parameter of `conditionalExpression` is mandatory.

Reply:
If `parentRightId` is `null`, it means `parentRight.node` is static and can be safely referenced. This case is covered in [assignment-static](https://github.com/babel/babel/blob/ac487d4a55cb70f96ac3e1019680943560a56f71/packages/babel-plugin-proposal-object-rest-spread/test/fixtures/object-rest/assignment-pattern-static/output.js).

If a `parentPath` is an assignment pattern, the `right` must presents, so `t.cloneNode(parentRight.node)` should always be truthy.

Maybe I can rename it into `nonStaticParentRightId`?

Reply:
I'm not sure I follow. If `parentRightId` is `null`, this will call `conditionalExpression` like this:
```js
t.conditionalExpression(/* objRefName !== void 0 */, /* objRefName */, null)
```
which isn't valid, because the third parameter can't be null.

Reply:
`parentRightId` is generated from `scope.maybeGenerateMemoised(parentRight.node)`, which is `null` when `t.isStatic(parentRight.node)` is true according to the [implementation](https://github.com/babel/babel/blob/fee4d936a3443c4d8812cac208e567e0ba8edc9d/packages/babel-traverse/src/scope/index.js#L330).

Reply:
Omg I was stupid :man_facepalming: I was reading `a ?? b` as `a == null ? null : b`

Reply:
(damn programming homework should also be blamed)